### PR TITLE
Allow searching for firmware using defined keywords

### PR DESCRIPTION
**Pull request by hughsie**
To migrate, do:

    FLASK_APP=app/__init__.py flask-2 modifydb

Fixes https://github.com/hughsie/lvfs-website/issues/71

### CODE REVIEW

File: app/templates/devicelist.html [hunk 1]
```diff
@@ -33,6 +33,14 @@ <h1>Devices using LVFS for firmware updates</h1>
...
         please see the <a href="/vendorlist">vendor engagement list</a>.
       </p>
+      <form method="GET" class="form-inline mb-3" action="/lvfs/search">
+        <div class="input-group">
+          <input type="text" class="form-control" aria-label="search" name="value" required>
+          <div class="input-group-append">
+            <button class="input-group-text" type="submit">Search</button>
+          </div>
+        </div>
+      </form>
```
Comment (superm1) [inline app/templates/devicelist.html]: Could you also keep this on the main home page?

File: app/views_component.py [hunk 2]
```diff
@@ -146,3 +169,65 @@ def firmware_requirement_add():
...
+    # log
+    flash('Removed keyword %s' % kw.value, 'info')
+    return redirect(url_for('.firmware_component_show',
+                            component_id=md.component_id,
+                            page='keywords'))
+
+@app.route('/lvfs/component/keyword/add', methods=['POST'])
+@login_required
+def firmware_keyword_add():
+    """ Modifies the update urgency and release notes for the update """
```
Comment (superm1) [inline app/views_component.py]: comment should be updated for what this is

File: app/templates/devicelist.html (hunk 1 above)
Comment (hughsie) [reply app/templates/devicelist.html]: I did originally have it there -- I moved it for two main reasons:

 * It doesn't work very well on a small screen (i.e. it forces the top bar into 2-3 "rows")
 * It's not really good enough yet

On the latter point you might notice I'm logging the search string and the number of results; after a few weeks deployed we can see how real users are using this and fix things accordingly. I think putting it so "front and center" it needs to be much better than it is now.

Comment (superm1) [reply app/templates/devicelist.html]: OK.

File: app/models.py [hunk 3]
```diff
@@ -294,6 +294,61 @@ def __init__(self, component_id=None, value=None):
...
+                 'for',
+                 'limited',
+                 'system',
+                 'the',
+                 'update']:
+        return False
+    return True
+
+def _sanitize_keyword(value):
+    for rpl in ['(', ')', '[', ']']:
```
Comment (superm1) [inline app/models.py]: You probably want to look for escape characters i'd think too right?

**Comment by hughsie** [comment]
Okay, merged for wider feedback. Check out https://fwupd.org/lvfs/search for the first cut.

**Comment by hughsie** [comment]
Some initial results:
screenshot from 2018-03-01 11-24-03

One thing that's obvious from the logs is that some people are searching for software, e.g. "Ubuntu" -- we should do something more useful there to help them.

**Pull request closed without merging**